### DISCUSSION TITLE
QueryBuilder doesn't insert default column values

### DISCUSSION THREAD

**Discussion by ramirofarias** [mikro-orm/mikro-orm#3214] [Q&A]
**Describe the bug**
When trying to insert values with query builder without specifying values that have defaults defined in the entity model, the data is not generated automatically and a "violates not-null constraint" is thrown.

```
//country.entity.ts
@Entity()
export class Country {
  @Property()
  name: string;
  @Property()
  code: string;
  @Property()
  phoneCode: number;
  @Property()
  createdAt: Date = new Date();
}

//DatabaseSeeder.ts
export class DatabaseSeeder extends Seeder {
  async run(em: EntityManager): Promise<void> {
    const countries = await this.generateCountries(em);
   }
 
   async generateCountries(em: EntityManager): Promise<Country[]> {
    const qb = em.createQueryBuilder(Country);
    qb.insert([
      { code: 'AF', name: 'Afghanistan', phoneCode: 93 },
      { code: 'AL', name: 'Albania', phoneCode: 355 },
    ]);

    const knex = qb.getKnexQuery();
    const res = await em.getConnection().execute(knex);
    const entities = res.map((c) => em.map(Country, c));
    return entities;
  }
 }
```

**Stack trace**
```
error: insert into "country" ("code", "name", "phone_code") values ('AF', 'Afghanistan', 93), ('AL', 'Albania', 355) returning "id" - null value in column "created_at" of relation "country" violates not-null constraint
    at Parser.parseErrorMessage (/home/rami/Projects/nestjs-mikro-orm/node_modules/pg-protocol/src/parser.ts:369:69)
    at Parser.handlePacket (/home/rami/Projects/nestjs-mikro-orm/node_modules/pg-protocol/src/parser.ts:188:21)
    at Parser.parse (/home/rami/Projects/nestjs-mikro-orm/node_modules/pg-protocol/src/parser.ts:103:30)
    at Socket.<anonymous> (/home/rami/Projects/nestjs-mikro-orm/node_modules/pg-protocol/src/index.ts:7:48)
    at Socket.emit (node:events:390:28)
    at Socket.emit (node:domain:475:12)
    at addChunk (node:internal/streams/readable:315:12)
    at readableAddChunk (node:internal/streams/readable:289:9)
    at Socket.Readable.push (node:internal/streams/readable:228:10)
    at TCP.onStreamRead (node:internal/stream_base_commons:199:23) {
  length: 236,
  severity: 'ERROR',
  code: '23502',
  detail: 'Failing row contains (1, null, null, Afghanistan, AF, 93).',
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: 'public',
  table: 'country',
  column: 'created_at',
  dataType: undefined,
  constraint: undefined,
  file: 'execMain.c',
  line: '1947',
  routine: 'ExecConstraints'
}
...
```

**To Reproduce**
1. Create a query builder for any entity with defined default values
2. Try to insert an array with said query builder in a DatabaseSeeder

**Expected behavior**
Default values are inserted without the need to define them

**Additional context**
I'm trying to use qb to insert a big array of Countries to seed a table.

**Versions**

| Dependency | Version |
| - | - |
| node | 16.13.1 |
| typescript | 4.5.4 |
| mikro-orm | 5.1.3 |
| your-driver | mysql - 5.1.3 |

**Comment by B4nan** [comment]
This is expected, you dont work with entity instances anywhere in your code (except the final `map` call which comes after the insert query is fired). With QB things need to be explicit or have database level defaults, property initializers are not used, as there are no entity instances created. You can either set database default for such column (e.g. `defaultRaw: 'now'`) or use UoW that will do multi insert query on its own automatically.

Why do you even use QB for this? UoW will do exactly the same as what you did with QB, but with less code (and automatic batching).

**Reply by ramirofarias** [reply]
Hi, thanks for the quick reply! I'm trying to use QB to insert a big array of objects (all countries) all at once instead of calling create on each of the elements.

Is there a way to do that with UoW? As far as I know there is no way to pass an array of entities to the create method. As a workaround, I created CountryFactory and I'm looping through the array and calling makeOne on each of the elements.

**Reply by B4nan** [reply]
This

```ts
const qb = em.createQueryBuilder(Country);
    qb.insert([
      { code: 'AF', name: 'Afghanistan', phoneCode: 93 },
      { code: 'AL', name: 'Albania', phoneCode: 355 },
    ]);

    const knex = qb.getKnexQuery();
    const res = await em.getConnection().execute(knex);
    const entities = res.map((c) => em.map(Country, c));
    return entities;
```

is equal to this

```ts
const data = [
  { code: 'AF', name: 'Afghanistan', phoneCode: 93 },
  { code: 'AL', name: 'Albania', phoneCode: 355 },
];
const countries = data.map(item => em.create(Country, item));
await em.persist(countries).flush();
return countries;
```

And in context of seeder I would say you dont even have to flush them, they should all get cascade persisted if you use them and the flush will happen eventually.

**Reply by B4nan** [reply]
Also instead of the manual execution and mapping, you can just do `qb.getResultList()` and it will return the entities for you.

**Reply by ramirofarias** [reply]
> This
> 
> ```ts
> const qb = em.createQueryBuilder(Country);
>     qb.insert([
>       { code: 'AF', name: 'Afghanistan', phoneCode: 93 },
>       { code: 'AL', name: 'Albania', phoneCode: 355 },
>     ]);
> 
>     const knex = qb.getKnexQuery();
>     const res = await em.getConnection().execute(knex);
>     const entities = res.map((c) => em.map(Country, c));
>     return entities;
> ```
> 
> is equal to this
> 
> ```ts
> const data = [
>   { code: 'AF', name: 'Afghanistan', phoneCode: 93 },
>   { code: 'AL', name: 'Albania', phoneCode: 355 },
> ];
> const countries = data.map(item => em.create(Country, item));
> await em.persist(countries).flush();
> return countries;
> ```
> 
> And in context of seeder I would say you dont even have to flush them, they should all get cascade persisted if you use them and the flush will happen eventually.

Nice, that's what I was doing but with a for in loop instead of map. Can I add this to the Seeding documentation as an example?

**Reply by B4nan** [reply]
Sure
